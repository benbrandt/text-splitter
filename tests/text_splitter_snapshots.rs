use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
