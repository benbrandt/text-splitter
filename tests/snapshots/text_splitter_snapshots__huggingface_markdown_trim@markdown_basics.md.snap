---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/markdown_basics.md
---
- "Markdown: Basics"
- "=========="
- "======"
- "<ul id=\"ProjectSubmenu"
- "\">"
- "<li><a href="
- "\"/projects/markdown/\" title="
- "\"Markdown Project Page\">Main</"
- a></li>
- "<li><a class=\"selected"
- "\" title=\"Markdown Basics\">"
- Basics</a></li
- ">"
- "<li><a href="
- "\"/projects/markdown/syntax\" title"
- "=\"Markdown Syntax"
- "Documentation\">Syntax<"
- /a></li>
- "<li><a href="
- "\"/projects/markdown/license\" title"
- "=\"Pricing and License Information\">"
- License</a></li>
- "<li><a href="
- "\"/projects/markdown/dingus\""
- "title=\"Online Markdown Web Form\">"
- Dingus</a></li
- ">\n</ul>"
- "Getting the Gist of Markdown'"
- s Formatting Syntax
- "----------"
- "----------"
- "----------"
- "----------"
- "--------"
- This page offers a brief overview of what
- "it's like to use Markdown.\nThe"
- "[syntax page] [s] provides complete,"
- detailed documentation for
- "every feature, but Markdown should be very easy"
- to pick up simply by
- looking at a few examples of it in action.
- The examples on this page
- "are written in a before/after style, showing"
- example syntax and the
- HTML output produced by Markdown.
- "It's also helpful to simply try Markdown"
- "out; the [Dingus] [d]"
- is a
- web application that allows you type your own Markdown
- "-formatted text"
- and translate it to XHTML.
- "**Note:** This document is itself"
- "written using Markdown; you\ncan ["
- "see the source for it by adding '.text"
- "' to the URL] [src]"
- "."
- "[s]: /projects/markdown/"
- "syntax  \"Markdown Syntax\""
- "[d]: /projects/markdown/"
- "dingus  \"Markdown Dingus\""
- "[src]: /projects/markdown"
- "/basics.text\n\n\n##"
- "Paragraphs, Headers, Blockquotes"
- "##"
- A paragraph is simply one or more consecutive lines of
- "text, separated"
- by one or more blank lines.
- (A blank line is any line that looks like
- a blank line -- a line containing nothing but
- spaces or tabs is
- considered blank.)
- Normal paragraphs should not be indented with
- spaces or tabs.
- "Markdown offers two styles of headers: *"
- Setext* and *atx*
- ".\nSetext-style headers for"
- "`<h1>` and `<"
- "h2>` are created by\n\"underlining"
- "\" with equal signs (`=`) and"
- "hyphens (`-`),"
- respectively.
- "To create an atx-style header, you"
- "put 1-6 hash marks (`#"
- "`) at the\nbeginning of the line --"
- the number of hashes equals the resulting
- HTML header level.
- Blockquotes are indicated using email-style
- "'`>`' angle brackets."
- "Markdown:"
- A First Level Header
- "=========="
- "=========="
- A Second Level Header
- "----------"
- "----------"
- "-"
- Now is the time for all good men to come
- to
- the aid of their country. This is just a
- regular paragraph.
- The quick brown fox jumped over the lazy
- "dog's back."
- "### Header 3"
- "> This is a blockquote.\n    >"
- "> This is the second paragraph in the"
- "blockquote.\n    >"
- "> ## This is an H2 in a"
- "blockquote\n\n\nOutput:"
- "<h1>A First Level Header<"
- /h1>
- "<h2>A Second Level Header<"
- /h2>
- "<p>Now is the time for all good"
- men to come to
- the aid of their country. This is just a
- regular paragraph.</p>
- "<p>The quick brown fox jumped over the"
- "lazy\n    dog's back.</p>"
- "<h3>Header 3</"
- "h3>\n\n    <blockquote>"
- "<p>This is a blockquote"
- ".</p>"
- "<p>This is the second paragraph in the"
- blockquote.</p>
- "<h2>This is an H2 in"
- a blockquote</h2>
- "</blockquote>\n\n\n\n###"
- "Phrase Emphasis ###"
- Markdown uses asterisks and
- underscores to indicate spans of emphasis.
- "Markdown:"
- Some of these words *are emphasized*.
- Some of these words _are emphasized also_.
- Use two asterisks for **strong
- emphasis**.
- "Or, if you prefer, __use two"
- "underscores instead__.\n\nOutput:"
- "<p>Some of these words <em>"
- are emphasized</em>.
- Some of these words <em>are emphasized also
- "</em>.</p>"
- "<p>Use two asterisks for"
- "<strong>strong emphasis</strong>."
- "Or, if you prefer, <strong>use"
- two underscores instead</strong>.
- "</p>\n\n\n\n## Lists ##"
- Unordered (bulleted) lists use
- "asterisks, pluses, and"
- "hyphens (`*`,"
- "`+`, and `-`) as"
- list markers. These three markers are
- "interchangable; this:"
- "*   Candy.\n    *   Gum."
- "*   Booze.\n\nthis:"
- "+   Candy.\n    +   Gum."
- "+   Booze.\n\nand this:"
- "-   Candy.\n    -   Gum."
- "-   Booze."
- "all produce the same output:"
- "<ul>"
- "<li>Candy.</li"
- ">"
- "<li>Gum.</"
- li>
- "<li>Booze.</"
- "li>\n    </ul>"
- "Ordered (numbered) lists use regular numbers,"
- "followed by periods, as\nlist markers:"
- "1.  Red\n    2.  Green\n    3.  Blue"
- "Output:"
- "<ol>"
- "<li>Red</li>"
- "<li>Green</li>"
- "<li>Blue</li>"
- "</ol>"
- "If you put blank lines between items,"
- "you'll get `<p>` tags"
- for the
- list item text.
- You can create multi-paragraph list items by
- indenting
- the paragraphs by 4 spaces or 1 tab
- ":"
- "*   A list item.\n\n        With multiple paragraphs."
- "*   Another item in the list.\n\nOutput:"
- "<ul>"
- "<li><p>A list item"
- ".</p>"
- "<p>With multiple paragraphs.</"
- p></li>
- "<li><p>Another item in"
- the list.</p></
- "li>\n    </ul>\n\n\n\n###"
- "Links ###"
- "Markdown supports two styles for creating links: *"
- "inline* and\n*reference*"
- "."
- "With both styles, you use square brackets to"
- delimit the
- text you want to turn into a link.
- Inline-style links use parentheses immediately after
- "the link text.\nFor example:"
- "This is an [example link](http:"
- //example.com/).
- "Output:"
- "<p>This is an <a"
- "href=\"http://"
- "example.com/\">"
- example link</a>.</p
- ">"
- "Optionally, you may include a title attribute"
- "in the parentheses:"
- "This is an [example link](http:"
- "//example.com/ \"With a Title"
- "\").\n\nOutput:"
- "<p>This is an <a"
- "href=\"http://"
- "example.com/\" title=\"With a"
- "Title\">"
- example link</a>.</p
- ">"
- Reference-style links allow you to refer to your
- "links by names, which"
- "you define elsewhere in your document:"
- "I get 10 times more traffic from [Google]"
- "[1] than from"
- "[Yahoo][2] or [MSN"
- "][3]."
- "[1]: http://"
- "google.com/        \"Google\""
- "[2]: http://"
- "search.yahoo.com/  \"Yahoo"
- "Search\""
- "[3]: http://"
- "search.msn.com/    \""
- "MSN Search\"\n\nOutput:"
- "<p>I get 10 times more traffic from"
- "<a href=\"http:/"
- "/google.com/\""
- "title=\"Google\">Google</a"
- "> than from <a href=\""
- "http://"
- "search.yahoo.com/\""
- "title=\"Yahoo Search\">Yahoo</"
- "a> or <a href=\""
- "http://"
- "search.msn.com/\""
- "title=\"MSN Search\">MSN"
- "</a>.</p>"
- The title attribute is optional.
- "Link names may contain letters,"
- "numbers and spaces, but are *not* case"
- "sensitive:"
- I start my morning with a cup of coffee and
- "[The New York Times][NY Times]"
- "."
- "[ny times]: http://"
- www.nytimes.com/
- "Output:"
- "<p>I start my morning with a cup"
- of coffee and
- "<a href=\"http:/"
- "/www.nytimes.com/\""
- ">The New York Times</a>."
- "</p>\n\n\n### Images"
- "###"
- Image syntax is very much like link syntax.
- "Inline (titles are optional):"
- "!["
- "alt text](/path/to/"
- "img.jpg \"Title\")"
- "Reference-style:"
- "![alt text][id]"
- "[id]: /path/to/"
- "img.jpg \"Title\""
- "Both of the above examples produce the same output:"
- "<img src=\"/path/"
- "to/img.jpg\""
- "alt=\"alt text\" title="
- "\"Title\" />\n\n\n\n### Code"
- "###"
- "In a regular paragraph, you can create code span"
- by wrapping text in
- backtick quotes. Any ampersands (
- "`&`) and angle brackets (`<"
- "` or\n`>`"
- ) will automatically be translated into HTML entities
- ". This makes"
- it easy to use Markdown to write about
- "HTML example code:"
- "I strongly recommend against using any `<blink>"
- "` tags."
- I wish SmartyPants used named entities like
- "`&mdash;`"
- "instead of decimal-encoded entities like `&"
- "#8212;`.\n\nOutput:"
- "<p>I strongly recommend against using any"
- "<code>&lt;blink&"
- gt;</code> tags.<
- /p>
- "<p>I wish SmartyPants used"
- named entities like
- "<code>&amp;mdash"
- ;</code> instead of decimal-
- encoded
- "entities like <code>&amp;#"
- 8212;</code>.</
- p>
- To specify an entire block of pre-formatted
- "code, indent every line of"
- the block by 4 spaces or 1 tab.
- "Just like with code spans, `&`,"
- "`<`,\nand `>`"
- "characters will be escaped automatically.\n\nMarkdown:"
- If you want your page to validate under
- "XHTML 1.0 Strict,"
- "you've got to put paragraph tags in your"
- "blockquotes:"
- "<blockquote>"
- "<p>For example.</p>"
- "</blockquote>\n\nOutput:"
- "<p>If you want your page to"
- validate under XHTML 1.0
- "Strict,"
- "you've got to put paragraph tags in your"
- "blockquotes:</p>"
- "<pre><code>&lt;"
- blockquote&gt;
- "&lt;p&gt;For"
- example.&lt;/p&
- gt;
- "&lt;/blockquote&"
- gt;
- "</code></pre>"

